### PR TITLE
Ensure NewStreamingMap and NewStreamingSet panic when input is not ordered

### LINKED ISSUE #3560: NewStreamingSet is not checking inputs are sorted correctly
It tries to (https://github.com/attic-labs/noms/blob/master/go/types/set.go#L50) but lastVal is never assigned to, so the check doesn't work.

### PR DESCRIPTION
**Pull request by willhite**
Fixes #3560

### CODE REVIEW

File: go/types/map.go [hunk 1]
```diff
@@ -38,34 +44,45 @@ func NewMap(kv ...Value) Map {
...
+// channel that will produce a StreamingMapResult. Values sent to the input
+// channel must be alternating keys and values. (e.g. k1, v1, k2, v2...).
+// Moreover keys must be non-nil and need to be added to the channel in Noms 
+// sortorder. If these constraints are violated, the result will contain a 
+// non-nil Error field and the map will only contain key-value pairs received on
+// the input channel before the first violation was found. Once the input channel
+// is closed by the caller, the StreamingMapResult will be sent to the output
+// channel. See graph_builder.go for building collections with values that are
+// not in order.
+func NewStreamingMap(vrw ValueReadWriter, kvs <-chan Value) <-chan StreamingMapResult {
```
Comment (aboodman) [inline go/types/map.go]: I don't think we should go to any trouble to allow callers to catch this condition gracefully. Just panic. If you want to test that case, then mock out the `go` statement or do something else to allow testing.

**Comment by willhite** [comment]
@aboodman PTAL redid this so as not to change api, overwrote last commit

**Comment by willhite** [comment]
@aboodman, I believe this is ready to go. It's just waiting on a PR approval.

**Comment by aboodman** [comment]
Can @arv or @rafael-atticlabs review this? I won't be able to get to it today or tomorrow.

**Comment by aboodman** [comment]
or @kalman

File: go/types/set.go [hunk 2]
```diff
@@ -38,24 +38,37 @@ func NewSet(v ...Value) Set {
...
+}
+
+func readSetInput(vrw ValueReadWriter, vChan <-chan Value, outChan chan<- Set) {
+	defer close(outChan)
+	ch := newEmptySetSequenceChunker(vrw, vrw)
+	var lastV Value
+	for v := range vChan {
+		d.PanicIfTrue(v == nil)
+		if lastV != nil {
+			d.PanicIfFalse(lastV == nil || lastV.Less(v))
```
Comment (willhite) [inline go/types/set.go]: done

File: go/types/set.go [hunk 3]
```diff
@@ -38,24 +38,37 @@ func NewSet(v ...Value) Set {
 // out of order will result in a panic. Once the input channel is closed
 // by the caller, a finished Set will be sent to the output channel. See
 // graph_builder.go for building collections with values that are not in order.
-func NewStreamingSet(vrw ValueReadWriter, vals <-chan Value) <-chan Set {
+func NewStreamingSet(vrw ValueReadWriter, vChan <-chan Value) <-chan Set {
+	return newStreamingSet(vrw, vChan, readSetInputAsync)
```
Comment (willhite) [inline go/types/set.go]: done

File: go/types/map.go [hunk 4]
```diff
@@ -46,30 +46,42 @@ func NewMap(kv ...Value) Map {
 // closed by the caller, a finished Map will be sent to the output channel. See
 // graph_builder.go for building collections with values that are not in order.
 func NewStreamingMap(vrw ValueReadWriter, kvs <-chan Value) <-chan Map {
-	var k Value
+	return newStreamingMap(vrw, kvs, readMapInputAsync)
```
Comment (willhite) [inline go/types/map.go]: done.

**Pull request merged**